File: Python/mergeSort.py
def merge(arr: list[int],s:int,e:int):
    left=list()
    right=list()
    mid=(s+e)//2
    leftLength=mid-s+1
    rightLength=e-mid
    arrayIndex=s
    for i in range(0,leftLength):
       left.append(arr[arrayIndex])
       arrayIndex+=1
    
    arrayIndex=mid+1
    for i in range(0,rightLength):
        right.append(arr[arrayIndex])
        arrayIndex+=1
    
    arrIndex=s
    leftIndex,rightIndex=0,0
    while(leftIndex<len(left) and rightIndex<len(right)):
        if(left[leftIndex]<right[rightIndex]):
            arr[arrIndex]=left[leftIndex]
            arrIndex+=1
            leftIndex+=1
        else:
            arr[arrIndex]=right[rightIndex]
            arrIndex+=1
            rightIndex+=1
        
    while(leftIndex<len(left)):
        arr[arrIndex]=left[leftIndex]
        arrIndex+=1
        leftIndex+=1
    while(rightIndex<len(right)):
        arr[arrIndex]=right[rightIndex]
        arrIndex+=1
        rightIndex+=1
        
    


def mergeSort(arr: list[int],s:int,e: int):
    if(s>=e):
        return
    mid=(s+e)//2
    mergeSort(arr,s,mid)
    mergeSort(arr,mid+1,e)
    merge(arr,s,e)

arr=list(map(int,input().split(' ')))
s=0
e=len(arr)-1
mergeSort(arr,s,e)
print(arr)

File: C++/merge-sort.cpp
#include <iostream>
#include <vector>

void merge(std::vector<int>& arr, int left, int middle, int right) {
    int n1 = middle - left + 1;
    int n2 = right - middle;

    // Create temporary arrays
    std::vector<int> left_arr(n1);
    std::vector<int> right_arr(n2);

    // Copy data to temporary arrays left_arr[] and right_arr[]
    for (int i = 0; i < n1; i++) {
        left_arr[i] = arr[left + i];
    }
    for (int i = 0; i < n2; i++) {
        right_arr[i] = arr[middle + 1 + i];
    }

    // Merge the two sorted arrays back into arr[]
    int i = 0; 
    int j = 0;
    int k = left; 

    while (i < n1 && j < n2) {
        if (left_arr[i] <= right_arr[j]) {
            arr[k] = left_arr[i];
            i++;
        } else {
            arr[k] = right_arr[j];
            j++;
        }
        k++;
    }

    // Copy the remaining elements of left_arr[], if any
    while (i < n1) {
        arr[k] = left_arr[i];
        i++;
        k++;
    }

    // Copy the remaining elements of right_arr[], if any
    while (j < n2) {
        arr[k] = right_arr[j];
        j++;
        k++;
    }
}

void merge_sort(std::vector<int>& arr, int left, int right) {
    if (left < right) {
        int middle = left + (right - left) / 2;

        // Sort first and second halves
        merge_sort(arr, left, middle);
        merge_sort(arr, middle + 1, right);

        // Merge the sorted halves
        merge(arr, left, middle, right);
    }
}

int main() {
    int n;
    std::cout << "Enter the number of elements: ";
    std::cin >> n;

    std::vector<int> input_list(n);
    std::cout << "Enter " << n << " elements:\n";
    for (int i = 0; i < n; i++) {
        std::cin >> input_list[i];
    }

    merge_sort(input_list, 0, n - 1);

    std::cout << "Sorted list: ";
    for (int i = 0; i < n; i++) {
        std::cout << input_list[i] << " ";
    }
    std::cout << std::endl;

    return 0;
}
